spark/src/main/scala/org/apache/hadoop/fs/versioned/VersionedFileSystem.scala

```diff
@@ -0,0 +1,138 @@
+    baseFS = FileSystem.get(baseURI, conf)
+    baseFS.initialize(baseURI, conf)
+  }
+
+  override def getUri: URI =
+    baseURI
+
+  override def open(f: Path, bufferSize: Int): FSDataInputStream =
+    baseFS.open(setScheme(baseURI.getScheme)(f), bufferSize)
+
+  override def create(
+      f: Path,
+      permission: FsPermission,
+      overwrite: Boolean,
+      bufferSize: Int,
+      replication: Short,
+      blockSize: Long,
+      progress: Progressable): FSDataOutputStream =
+    baseFS.create(baseVersionedPath(f), permission, overwrite, bufferSize, replication, blockSize, progress)
+
+  override def append(f: Path, bufferSize: Int, progress: Progressable): FSDataOutputStream =
+    baseFS.append(baseVersionedPath(f), bufferSize, progress)
+
+  override def rename(src: Path, dst: Path): Boolean =
+    baseFS.rename(baseVersionedPath(src), baseVersionedPath(dst))
+
+  override def delete(f: Path, recursive: Boolean): Boolean =
+    baseFS.delete(baseVersionedPath(f), recursive)
+
+  override def listStatus(f: Path): Array[FileStatus] =
+    baseFS.listStatus(setScheme(baseURI.getScheme)(f)) map { fileStatus =>
+      new FileStatus(
+        fileStatus.getLen,
+        fileStatus.isDirectory,
+        fileStatus.getReplication.toInt,
+        fileStatus.getBlockSize,
+        fileStatus.getModificationTime,
+        versionedPath(fileStatus.getPath)
+      )
+    } filter { fileStatus =>
+      fileStatus.getPath.toUri.getSchemeSpecificPart.contains(version) ||
+      fileStatus.isDirectory
+    }
+
+  override def setWorkingDirectory(new_dir: Path): Unit =
+    baseFS.setWorkingDirectory(baseVersionedPath(new_dir))
+
+  override def getWorkingDirectory: Path =
+    setScheme(baseURI.getScheme)(baseFS.getWorkingDirectory)
+
+  override def mkdirs(f: Path, permission: FsPermission): Boolean =
+    baseFS.mkdirs(baseVersionedPath(f), permission)
+
+  override def getFileStatus(f: Path): FileStatus = {
+    val fileStatus = baseFS.getFileStatus(setScheme(baseURI.getScheme)(f))
+
+    new FileStatus(
+      fileStatus.getLen,
+      fileStatus.isDirectory,
+      fileStatus.getReplication.toInt,
+      fileStatus.getBlockSize,
+      fileStatus.getModificationTime,
+      versionedPath(fileStatus.getPath)
+    )
+  }
+
+  private lazy val appendVersion: Path => Path = path => {
+    val uri = path.toUri
+
+    val sp = mappings.foldLeft(uri.getSchemeSpecificPart) {
+      case (acc, (partition, versionedPartition)) if !acc.contains(versionedPartition) && acc.contains(partition) =>
+        acc.replace(partition, versionedPartition)
+      case (acc, _) => acc
+    }
+
+    new Path(new URI(s"${uri.getScheme}:$sp"))
+  }
+
+  private lazy val setScheme: String => Path => Path =
+    scheme => path => new Path(new URI(s"$scheme:${path.toUri.getSchemeSpecificPart}"))
+
+  private lazy val toSchemeAndVersion: String => Path => Path = scheme => setScheme(scheme) andThen appendVersion
```

Comment:
Any particular reason to use this function syntax here? And why are these values `lazy`?

Reply:
`lazy` is possibly a tad of an premature optimisation. The reason was this FS is not cached, so until any of those methods is called, it'd create the said FS faster.

As for the style, I quite like this haskell'ish syntax

Reply:
I doubt very much the time taken to initialise a `val` that points to a function is noticeable compared to loading a filesystem object and writing a Spark dataset through it. It might even be slower (also in a very insignificant way of course), as `lazy` incurs a synchronisation overhead...

As for Haskell syntax, the rest of the code base doesn't use it so I'd rather change it.

Reply:
I find this style encourages function composition (which it does here on a very small scale). If that's not FP... ;-) 

I guess i'll go with the consensus as i don't mind too much either way, just trying to be more FP and break away from ld java habits :). @pvighi @desbo, what's your preference ?

Reply:
I tend to avoid a having too strong of a personal preference in favour of consistency with the rest of the codebase

Reply:
Changed to defs